stl/inc/valarray

```diff
@@ -1372,6 +1372,11 @@ public:
     _NODISCARD size_t stride() const noexcept /* strengthened */ {
         return _Stride;
     }
+#if _HAS_CXX200
+    friend bool operator==(const slice& x, const slice& y) {

```

Comment:
We do a few things that go beyond the Standardese. One of them is marking functions with the `[[nodiscard]]` attribute, which warns about discarded return values (here, it would warn about a statement `slice1 == slice2;` when the user calls `operator==` but does nothing with the returned `bool` - perhaps they meant to test it with `if`, or perhaps they meant to write assignment instead of equality). We have a set of criteria for marking functions as `[[nodiscard]]`, and the most common criterion is "pure observer" - a function that has no side effects and is called only for its return value, which applies to all comparison operators.

Note that we currently apply this attribute via an internal macro, `_NODISCARD` (which is defined in a vcruntime header). This allows us to control it via an escape hatch (which has never been necessary) and allowed us to make it activated for C++17 and above (but we eventually got compiler support for C++14 mode). In the future we may use the `[[nodiscard]]` attribute directly in headers, but for now we should keep using the `_NODISCARD` macro for consistency.